### PR TITLE
Deduplicate environment objects in the graph

### PR DESCRIPTION
**Pull request by gorakong**
### Test Plan

BundleGraph sizes:

**Kitchen-Sink** (tiny app)
Before:
202885 bytes

After:
192659 bytes

---------------

**On a large app**
Before:
67503048 bytes

After:
64734182 bytes

**difference of 2.77 MB**

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.89s | +58.00ms |
| Cached | 563.00ms | -1.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 546.00ms | +31.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 73.00ms | +18.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 50.00ms | -6.00ms 🚀 |
| dist/modern/index.f81b5b4b.js | 1.11kb | +0.00b | 74.00ms | +18.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 77.00ms | +21.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 74.00ms | +18.00ms ⚠️ |
| dist/legacy/index.5859ab37.css | 77.00b | +0.00b | 75.00ms | +18.00ms ⚠️ |
| dist/modern/index.77ce15d4.css | 77.00b | +0.00b | 74.00ms | +18.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 19.72s | +18.00ms |
| Cached | 8.50s | +106.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.1964b8bb.js | 4.20kb | +0.00b | 894.00ms | +169.00ms ⚠️ |
| dist/UserProfile.f5fa9f60.js | 1.76kb | +0.00b | 894.00ms | +169.00ms ⚠️ |
| dist/NotFound.f4adce8d.js | 530.00b | +0.00b | 223.00ms | -15.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 485.62kb | +0.00b | 76.00ms | +7.00ms ⚠️ |
| dist/PermalinkedComment.1964b8bb.js | 4.20kb | +0.00b | 77.00ms | +8.00ms ⚠️ |
| dist/UserProfile.f5fa9f60.js | 1.76kb | +0.00b | 77.00ms | +8.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 58.00ms | +7.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.90m | -2.13s |
| Cached | 3.00s | +83.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/pdfRenderer.1c87caa3.js | 1.11mb | +0.00b | 37.60s | -3.05s 🚀 |
| dist/editorView.94307fd1.js | 587.53kb | +0.00b | 51.80s | -3.90s 🚀 |
| dist/EmojiPickerComponent.b302072e.js | 139.01kb | +0.00b | 9.06s | -1.57s 🚀 |
| dist/Toolbar.453ced4a.js | 99.77kb | +0.00b | 13.24s | -3.09s 🚀 |
| dist/media-viewer.b4647ff4.js | 72.87kb | +0.00b | 10.52s | -582.00ms 🚀 |
| dist/card.92b0f4f9.js | 54.02kb | +0.00b | 7.90s | -2.58s 🚀 |
| dist/card.43bef276.js | 51.77kb | +0.00b | 15.37s | +1.53s ⚠️ |
| dist/Modal.507a0880.js | 39.18kb | +0.00b | 3.72s | -778.00ms 🚀 |
| dist/component.e862677c.js | 30.82kb | +0.00b | 3.75s | +1.87s ⚠️ |
| dist/esm.572878ca.js | 27.67kb | +0.00b | 6.51s | -1.55s 🚀 |
| dist/component.c1c3a8ff.js | 22.54kb | +0.00b | 3.60s | -896.00ms 🚀 |
| dist/DatePicker.64b5aa93.js | 21.20kb | +0.00b | 6.12s | -2.99s 🚀 |
| dist/smartMediaEditor.ab51c6dd.js | 16.85kb | +0.00b | 11.50s | -4.70s 🚀 |
| dist/js.63570256.js | 16.55kb | +0.00b | 3.60s | -882.00ms 🚀 |
| dist/dropzone.491f3481.js | 16.06kb | +0.00b | 10.61s | -1.46s 🚀 |
| dist/ui.bfd5ec70.js | 14.21kb | +0.00b | 7.81s | -2.28s 🚀 |
| dist/workerHasher.c27b76fd.js | 11.90kb | +0.00b | 8.72s | +674.00ms ⚠️ |
| dist/component.dd6e8cc9.js | 6.20kb | +0.00b | 4.01s | +906.00ms ⚠️ |
| dist/card.89465567.js | 5.71kb | +0.00b | 9.43s | -1.66s 🚀 |
| dist/media-viewer.114c1cc8.js | 4.01kb | +0.00b | 10.52s | -576.00ms 🚀 |
| dist/EmojiPickerComponent.5d345edd.js | 3.62kb | +0.00b | 7.56s | -1.42s 🚀 |
| dist/png-chunks-extract.8c8d635d.js | 3.55kb | +0.00b | 3.60s | -887.00ms 🚀 |
| dist/index.10f02a5e.css | 3.46kb | +0.00b | 37.51s | -3.14s 🚀 |
| dist/dropzone.c3746b5e.js | 3.44kb | +0.00b | 10.30s | -1000.00ms 🚀 |
| dist/Modal.19c70720.js | 3.14kb | +0.00b | 991.00ms | -1.79s 🚀 |
| dist/clipboard.84ebdbd2.js | 2.97kb | +0.00b | 15.15s | +2.09s ⚠️ |
| dist/16.624bbe22.js | 2.49kb | +0.00b | 992.00ms | -857.00ms 🚀 |
| dist/card.4188bb6a.js | 2.15kb | +0.00b | 9.99s | -1.21s 🚀 |
| dist/date.57372c70.js | 1.96kb | +0.00b | 4.22s | -1.80s 🚀 |
| dist/images.c36248a4.js | 1.90kb | +0.00b | 4.33s | -489.00ms 🚀 |
| dist/feedback.80c8a178.js | 1.86kb | +0.00b | 5.21s | -3.87s 🚀 |
| dist/16.342e69d2.js | 1.86kb | +0.00b | 3.98s | -532.00ms 🚀 |
| dist/browser.00e50e54.js | 1.82kb | +0.00b | 15.17s | +1.66s ⚠️ |
| dist/16.d173f3df.js | 1.79kb | +0.00b | 3.77s | +487.00ms ⚠️ |
| dist/workerHasher.f1855576.js | 1.75kb | +0.00b | 6.47s | -1.53s 🚀 |
| dist/workerHasher.cbff2c2b.js | 1.75kb | +0.00b | 11.43s | -4.11s 🚀 |
| dist/list-number.9bb92c49.js | 1.68kb | +0.00b | 4.38s | -581.00ms 🚀 |
| dist/status.acecc9f0.js | 1.68kb | +0.00b | 4.53s | -555.00ms 🚀 |
| dist/code.6d1ec8ce.js | 1.61kb | +0.00b | 4.18s | -1.84s 🚀 |
| dist/heading6.f14573d2.js | 1.53kb | +0.00b | 4.63s | -4.44s 🚀 |
| dist/heading3.dede5b4c.js | 1.51kb | +0.00b | 4.61s | -573.00ms 🚀 |
| dist/16.adfefbb7.js | 1.46kb | +0.00b | 3.79s | +501.00ms ⚠️ |
| dist/emoji.5072e4af.js | 1.45kb | +0.00b | 4.30s | -517.00ms 🚀 |
| dist/16.186663a4.js | 1.45kb | +0.00b | 991.00ms | -858.00ms 🚀 |
| dist/16.2c3c2005.js | 1.44kb | +0.00b | 4.15s | -462.00ms 🚀 |
| dist/16.45c1cffe.js | 1.40kb | +0.00b | 4.14s | -465.00ms 🚀 |
| dist/heading5.fe670803.js | 1.40kb | +0.00b | 4.63s | -3.17s 🚀 |
| dist/expand.739e3b88.js | 1.38kb | +0.00b | 6.11s | -2.96s 🚀 |
| dist/16.8ff06a1d.js | 1.36kb | +0.00b | 3.98s | -533.00ms 🚀 |
| dist/16.263c691c.js | 1.34kb | +0.00b | 992.00ms | -857.00ms 🚀 |
| dist/heading2.5c4ee285.js | 1.33kb | +0.00b | 4.56s | -619.00ms 🚀 |
| dist/16.b7512655.js | 1.32kb | +0.00b | 4.01s | -1.08s 🚀 |
| dist/16.262b91f7.js | 1.31kb | +0.00b | 4.15s | -465.00ms 🚀 |
| dist/mention.76dc783a.js | 1.29kb | +0.00b | 4.41s | -548.00ms 🚀 |
| dist/heading4.8bbff8c5.js | 1.29kb | +0.00b | 4.61s | -600.00ms 🚀 |
| dist/Modal.d58fb436.js | 1.28kb | +0.00b | 3.60s | -887.00ms 🚀 |
| dist/16.436e2680.js | 1.27kb | +0.00b | 992.00ms | -857.00ms 🚀 |
| dist/16.eaa9cadb.js | 1.27kb | +0.00b | 3.98s | -536.00ms 🚀 |
| dist/16.19a76383.js | 1.26kb | +0.00b | 4.21s | +234.00ms ⚠️ |
| dist/16.62eaafc3.js | 1.26kb | +0.00b | 3.79s | +502.00ms ⚠️ |
| dist/divider.2e62a51b.js | 1.25kb | +0.00b | 4.30s | -518.00ms 🚀 |
| dist/quote.487996a9.js | 1.25kb | +0.00b | 4.50s | -566.00ms 🚀 |
| dist/component.3edf7abd.js | 1.23kb | +0.00b | 3.68s | +900.00ms ⚠️ |
| dist/16.0629cb86.js | 1.23kb | +0.00b | 3.69s | +914.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.3d09fdec.js | 1.23kb | +0.00b | 9.33s | -1.20s 🚀 |
| dist/decision.cdcb6e3b.js | 1.21kb | +0.00b | 4.30s | -1.79s 🚀 |
| dist/panel-warning.8b68f633.js | 1.21kb | +0.00b | 4.47s | -597.00ms 🚀 |
| dist/16.cd483156.js | 1.18kb | +0.00b | 4.11s | -503.00ms 🚀 |
| dist/heading1.a939a2da.js | 1.18kb | +0.00b | 4.53s | -648.00ms 🚀 |
| dist/panel-error.210bc7a0.js | 1.11kb | +0.00b | 4.42s | -555.00ms 🚀 |
| dist/panel.8b5546b1.js | 1.10kb | +0.00b | 4.50s | -566.00ms 🚀 |
| dist/table.4e67906a.js | 1.09kb | +0.00b | 5.55s | +374.00ms ⚠️ |
| dist/panel-success.a2e1eac9.js | 1.05kb | +0.00b | 4.45s | -620.00ms 🚀 |
| dist/panel-note.8e008c33.js | 1.05kb | +0.00b | 4.42s | -576.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.a20da0c5.js | 1003.00b | +0.00b | 10.60s | -800.00ms 🚀 |
| dist/media-card-analytics-error-boundary.ef6306e0.js | 999.00b | +0.00b | 7.90s | -2.61s 🚀 |
| dist/media-card-analytics-error-boundary.889a764d.js | 999.00b | +0.00b | 11.02s | -4.48s 🚀 |
| dist/media-viewer-analytics-error-boundary.880deb8d.js | 995.00b | +0.00b | 9.95s | -674.00ms 🚀 |
| dist/simpleHasher.b4d19bac.js | 755.00b | +0.00b | 11.50s | -4.11s 🚀 |
| dist/index.html | 119.00b | +0.00b | 2.83s | +905.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.7b715b4e.js | 2.43mb | +0.00b | 542.00ms | -79.00ms 🚀 |
| dist/pdfRenderer.1c87caa3.js | 1.11mb | +0.00b | 611.00ms | -62.00ms 🚀 |
| dist/editorView.94307fd1.js | 587.53kb | +0.00b | 617.00ms | -76.00ms 🚀 |
| dist/popup.8b8992f9.js | 169.17kb | +0.00b | 611.00ms | -77.00ms 🚀 |
| dist/EmojiPickerComponent.b302072e.js | 139.01kb | +0.00b | 581.00ms | -66.00ms 🚀 |
| dist/Toolbar.c11cfac5.js | 99.68kb | +0.00b | 617.00ms | -76.00ms 🚀 |
| dist/media-viewer.e8181fcd.js | 72.78kb | +0.00b | 600.00ms | -73.00ms 🚀 |
| dist/card.793062ce.js | 53.93kb | +0.00b | 599.00ms | -61.00ms 🚀 |
| dist/card.8b7fed58.js | 51.68kb | +0.00b | 617.00ms | -61.00ms 🚀 |
| dist/esm.572878ca.js | 27.67kb | +0.00b | 581.00ms | -91.00ms 🚀 |
| dist/component.c1c3a8ff.js | 22.54kb | +0.00b | 495.00ms | -56.00ms 🚀 |
| dist/DatePicker.2e53d84e.js | 21.12kb | +0.00b | 581.00ms | -65.00ms 🚀 |
| dist/smartMediaEditor.60d9f6fb.js | 16.76kb | +0.00b | 617.00ms | -76.00ms 🚀 |
| dist/js.63570256.js | 16.55kb | +0.00b | 495.00ms | -52.00ms 🚀 |
| dist/dropzone.45bc40ce.js | 16.00kb | +0.00b | 611.00ms | -77.00ms 🚀 |
| dist/ui.bfd5ec70.js | 14.21kb | +0.00b | 560.00ms | -100.00ms 🚀 |
| dist/workerHasher.c27b76fd.js | 11.90kb | +0.00b | 559.00ms | -113.00ms 🚀 |
| dist/component.dd6e8cc9.js | 6.20kb | +0.00b | 495.00ms | -56.00ms 🚀 |
| dist/card.89465567.js | 5.71kb | +0.00b | 599.00ms | -74.00ms 🚀 |
| dist/media-viewer.114c1cc8.js | 4.01kb | +0.00b | 611.00ms | -62.00ms 🚀 |
| dist/png-chunks-extract.8c8d635d.js | 3.55kb | +0.00b | 495.00ms | -52.00ms 🚀 |
| dist/EmojiPickerComponent.5f66d6ae.js | 3.53kb | +0.00b | 559.00ms | -114.00ms 🚀 |
| dist/index.10f02a5e.css | 3.46kb | +0.00b | 619.00ms | -77.00ms 🚀 |
| dist/dropzone.c3746b5e.js | 3.44kb | +0.00b | 610.00ms | -78.00ms 🚀 |
| dist/Modal.19c70720.js | 3.14kb | +0.00b | 485.00ms | -158.00ms 🚀 |
| dist/clipboard.84ebdbd2.js | 2.97kb | +0.00b | 611.00ms | -67.00ms 🚀 |
| dist/16.624bbe22.js | 2.49kb | +0.00b | 485.00ms | -51.00ms 🚀 |
| dist/ResourcedEmojiComponent.470ee09d.js | 2.15kb | +0.00b | 560.00ms | -100.00ms 🚀 |
| dist/card.4188bb6a.js | 2.15kb | +0.00b | 611.00ms | -77.00ms 🚀 |
| dist/date.57372c70.js | 1.96kb | +0.00b | 523.00ms | -67.00ms 🚀 |
| dist/images.c36248a4.js | 1.90kb | +0.00b | 527.00ms | -70.00ms 🚀 |
| dist/feedback.80c8a178.js | 1.86kb | +0.00b | 550.00ms | -101.00ms 🚀 |
| dist/16.342e69d2.js | 1.86kb | +0.00b | 514.00ms | -57.00ms 🚀 |
| dist/browser.00e50e54.js | 1.82kb | +0.00b | 611.00ms | -67.00ms 🚀 |
| dist/16.d173f3df.js | 1.79kb | +0.00b | 502.00ms | -49.00ms 🚀 |
| dist/workerHasher.f1855576.js | 1.75kb | +0.00b | 581.00ms | -91.00ms 🚀 |
| dist/workerHasher.cbff2c2b.js | 1.75kb | +0.00b | 617.00ms | -61.00ms 🚀 |
| dist/list-number.9bb92c49.js | 1.68kb | +0.00b | 531.00ms | -75.00ms 🚀 |
| dist/status.acecc9f0.js | 1.68kb | +0.00b | 543.00ms | -76.00ms 🚀 |
| dist/code.6d1ec8ce.js | 1.61kb | +0.00b | 523.00ms | -67.00ms 🚀 |
| dist/link.cda9435d.js | 1.53kb | +0.00b | 530.00ms | -75.00ms 🚀 |
| dist/heading6.f14573d2.js | 1.53kb | +0.00b | 550.00ms | -101.00ms 🚀 |
| dist/heading3.dede5b4c.js | 1.51kb | +0.00b | 550.00ms | -101.00ms 🚀 |
| dist/16.9e858e4a.js | 1.51kb | +0.00b | 513.00ms | -49.00ms 🚀 |
| dist/16.adfefbb7.js | 1.46kb | +0.00b | 501.00ms | -50.00ms 🚀 |
| dist/16.23a6fba4.js | 1.46kb | +0.00b | 502.00ms | -59.00ms 🚀 |
| dist/emoji.5072e4af.js | 1.45kb | +0.00b | 527.00ms | -70.00ms 🚀 |
| dist/16.186663a4.js | 1.45kb | +0.00b | 485.00ms | -51.00ms 🚀 |
| dist/16.2c3c2005.js | 1.44kb | +0.00b | 518.00ms | -71.00ms 🚀 |
| dist/16.45c1cffe.js | 1.40kb | +0.00b | 518.00ms | -54.00ms 🚀 |
| dist/heading5.fe670803.js | 1.40kb | +0.00b | 550.00ms | -101.00ms 🚀 |
| dist/expand.739e3b88.js | 1.38kb | +0.00b | 580.00ms | -66.00ms 🚀 |
| dist/16.8ff06a1d.js | 1.36kb | +0.00b | 518.00ms | -54.00ms 🚀 |
| dist/16.263c691c.js | 1.34kb | +0.00b | 485.00ms | -51.00ms 🚀 |
| dist/heading2.5c4ee285.js | 1.33kb | +0.00b | 543.00ms | -76.00ms 🚀 |
| dist/16.b7512655.js | 1.32kb | +0.00b | 518.00ms | -54.00ms 🚀 |
| dist/16.262b91f7.js | 1.31kb | +0.00b | 523.00ms | -67.00ms 🚀 |
| dist/mention.76dc783a.js | 1.29kb | +0.00b | 531.00ms | -75.00ms 🚀 |
| dist/heading4.8bbff8c5.js | 1.29kb | +0.00b | 550.00ms | -101.00ms 🚀 |
| dist/Modal.d58fb436.js | 1.28kb | +0.00b | 495.00ms | -52.00ms 🚀 |
| dist/16.436e2680.js | 1.27kb | +0.00b | 485.00ms | -51.00ms 🚀 |
| dist/layout.d9b11ad1.js | 1.27kb | +0.00b | 527.00ms | -70.00ms 🚀 |
| dist/16.eaa9cadb.js | 1.27kb | +0.00b | 513.00ms | -49.00ms 🚀 |
| dist/16.06e8a68f.js | 1.26kb | +0.00b | 502.00ms | -144.00ms 🚀 |
| dist/16.19a76383.js | 1.26kb | +0.00b | 513.00ms | -49.00ms 🚀 |
| dist/16.e5432caf.js | 1.26kb | +0.00b | 513.00ms | -49.00ms 🚀 |
| dist/16.62eaafc3.js | 1.26kb | +0.00b | 501.00ms | -50.00ms 🚀 |
| dist/divider.2e62a51b.js | 1.25kb | +0.00b | 527.00ms | -70.00ms 🚀 |
| dist/quote.487996a9.js | 1.25kb | +0.00b | 543.00ms | -75.00ms 🚀 |
| dist/component.3edf7abd.js | 1.23kb | +0.00b | 542.00ms | -79.00ms 🚀 |
| dist/action.8161ba48.js | 1.23kb | +0.00b | 523.00ms | -67.00ms 🚀 |
| dist/16.0629cb86.js | 1.23kb | +0.00b | 542.00ms | -79.00ms 🚀 |
| dist/media-card-analytics-error-boundary.3d09fdec.js | 1.23kb | +0.00b | 599.00ms | -61.00ms 🚀 |
| dist/decision.cdcb6e3b.js | 1.21kb | +0.00b | 526.00ms | -71.00ms 🚀 |
| dist/panel-warning.8b68f633.js | 1.21kb | +0.00b | 536.00ms | -77.00ms 🚀 |
| dist/16.cd483156.js | 1.18kb | +0.00b | 518.00ms | -54.00ms 🚀 |
| dist/list.e0e1a7a9.js | 1.18kb | +0.00b | 531.00ms | -75.00ms 🚀 |
| dist/heading1.a939a2da.js | 1.18kb | +0.00b | 543.00ms | -76.00ms 🚀 |
| dist/panel-error.210bc7a0.js | 1.11kb | +0.00b | 535.00ms | -71.00ms 🚀 |
| dist/panel.8b5546b1.js | 1.10kb | +0.00b | 536.00ms | -77.00ms 🚀 |
| dist/table.4e67906a.js | 1.09kb | +0.00b | 543.00ms | -76.00ms 🚀 |
| dist/panel-success.a2e1eac9.js | 1.05kb | +0.00b | 536.00ms | -77.00ms 🚀 |
| dist/panel-note.8e008c33.js | 1.05kb | +0.00b | 536.00ms | -77.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.a20da0c5.js | 1003.00b | +0.00b | 611.00ms | -77.00ms 🚀 |
| dist/media-card-analytics-error-boundary.ef6306e0.js | 999.00b | +0.00b | 599.00ms | -61.00ms 🚀 |
| dist/media-card-analytics-error-boundary.889a764d.js | 999.00b | +0.00b | 617.00ms | -61.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.880deb8d.js | 995.00b | +0.00b | 611.00ms | -62.00ms 🚀 |
| dist/simpleHasher.d91ba9fe.js | 755.00b | +0.00b | 559.00ms | -114.00ms 🚀 |
| dist/simpleHasher.b4d19bac.js | 755.00b | +0.00b | 617.00ms | -76.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 468.00ms | -84.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 13.96s | -23.00ms |
| Cached | 610.00ms | -4.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.59kb | +0.00b | 110.00ms | +7.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/287822194758648333)**

File: packages/core/core/src/Environment.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ const DEFAULT_ENGINES = {
   node: '>= 8.0.0',
 };
 
+export const envCache: Map<string, Environment> = new Map();
```
Comment (wbinnssmith) [inline packages/core/core/src/Environment.js:11]: these entries currently have no way of being freed 😕

File: packages/core/core/src/requests/AssetGraphRequest.js [hunk 2]
```diff
@@ -37,6 +37,7 @@ import ThrowableDiagnostic from '@parcel/diagnostic';
 import AssetGraph from '../AssetGraph';
 import {PARCEL_VERSION} from '../constants';
 import createEntryRequest from './EntryRequest';
+import {envCache} from '../Environment';
```
Comment (wbinnssmith) [inline packages/core/core/src/requests/AssetGraphRequest.js:40]: feels a bit leaky to export the entire cache map. maybe something like `getEnvironmentById`?

File: packages/core/core/src/requests/AssetGraphRequest.js [hunk 3]
```diff
@@ -201,6 +202,20 @@ export class AssetGraphBuilder {
 
     dumpToGraphViz(this.assetGraph, 'AssetGraph');
 
+    for (let asset of this.assetGraph.nodes.entries()) {
```
Comment (wbinnssmith) [inline packages/core/core/src/requests/AssetGraphRequest.js:205]: I don't think we should be iterating every node in the graph to do this. Can environments be normalized as they're created?

File: packages/core/core/src/AssetGraph.js [hunk 4]
```diff
@@ -127,6 +130,18 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
     };
   }
 
+  getOrSetEnvironment(input: Asset | Dependency | AssetGroup) {
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js:133]: why is this called `get` if it doesn't return anything?

Comment (wbinnssmith) [reply packages/core/core/src/AssetGraph.js:133]: maybe something like `normalizeEnvironment`?

**Review by devongovett — APPROVED** [review]
Looks good aside from the naming thing Will noted

**Pull request merged**